orttraining/orttraining/training_ops/cuda/optimizer/lamb.cu

```diff
@@ -519,11 +572,34 @@ void LambMultiTensorReductionFunctor<TIn1, TIn2, TOut1, TOut2, TBuf>::operator()
   ORT_ENFORCE(thread_count % GPU_WARP_SIZE == 0);
   ORT_ENFORCE((thread_count & (thread_count - 1)) == 0);
 
-  LambMultiTensorReductionImpl<TIn1, TIn2, TOut1, TOut2, TBuf><<<chunk_group.chunk_count, thread_count, shared_memory_size>>>(chunk_group);
+  const int num_blocks = chunk_group.chunk_count;
+  size_t w_buffer_size = num_blocks * sizeof(TOut1);
+  size_t d_buffer_size = num_blocks * sizeof(TOut2);
+
+  ORT_ENFORCE(w_buffer_size + d_buffer_size <= reduction_buffer_size);
+
+  TOut1 *w_buffer = reinterpret_cast<TOut1*>(reduction_buffer);
+  TOut2 *d_buffer = reinterpret_cast<TOut2*>(w_buffer + num_blocks);
+
+  // sync_range_and_lock is a struct consisting of (start_block, num_blocks, lock) for each tensor
+  // Note: Adding such info to chunk group causes overflow (unless max tensors is reduced)
+  const int max_tensors = ChunkGroup<4>::max_tensor_group_count;
+  LambMultiTensorSyncRangeAndLock initial = {0, 0, 0};
+  CudaKernel::CudaAsyncBuffer<LambMultiTensorSyncRangeAndLock> sync_range_and_lock(&kernel, initial, max_tensors);
+  for (int block_index = num_blocks-1; block_index >= 0; block_index--) {
+    int tensor_index = chunk_group.block_index_to_tensor_group_index[block_index];
+    auto& tensor_block_span = sync_range_and_lock.CpuPtr()[tensor_index];
+    tensor_block_span.leading_block = block_index;
+    tensor_block_span.number_blocks++;
+  }
+  sync_range_and_lock.CopyToGpu();
```

Comment:
Let's wrap this place in a new function such as
```
CudaKernel::CudaAsyncBuffer<LambMultiTensorSyncRangeAndLock> compute_tensor_range_and_lock(...)
```

Reply:
Good idea.